cc/mcts_player.cc

```diff
@@ -279,62 +279,60 @@ bool MctsPlayer::PlayMove(Coord c) {
   }
 
   // Handle consecutive passing or termination by move limit.
-  if (root_->game_over() || root_->at_move_limit()) {
-    float score = root_->position.CalculateScore(options_.komi);
-    result_string_ = FormatScore(score);
-    result_ = score < 0 ? -1 : score > 0 ? 1 : 0;
+  if (game != nullptr) {
+    if (root_->game_over() || root_->at_move_limit()) {
+      game->SetGameOverBecauseOfPasses(
+          root_->position.CalculateScore(options_.game_options.komi));
+    }
   }
 
   return true;
 }
 
-std::string MctsPlayer::FormatScore(float score) const {
-  return absl::StrFormat("%c+%.1f", score > 0 ? 'B' : 'W', std::abs(score));
-}
-
-void MctsPlayer::PushHistory(Coord c) {
-  history_.emplace_back();
-  History& history = history_.back();
-  history.c = c;
-  history.comment = root_->Describe();
-  history.node = root_;
-
+void MctsPlayer::UpdateGame(Coord c, Game* game) {
+  // Record which model(s) were used when running tree search for this move.
+  std::vector<std::string> models;
   if (!inferences_.empty()) {
-    // Record which model(s) were used when running tree search for this move.
-    std::vector<std::string> models;
     for (auto it = inferences_.rbegin(); it != inferences_.rend(); ++it) {
       if (it->last_move < root_->position.n()) {
         break;
       }
       models.push_back(it->model);
     }
     std::reverse(models.begin(), models.end());
-    auto model_comment = absl::StrCat("models:", absl::StrJoin(models, ","));
-    history.comment = absl::StrCat(model_comment, "\n", history.comment);
-    if (options_.verbose) {
-      MG_LOG(INFO) << model_comment;
-    }
+  }
+
+  // Build a comment for the move.
+  auto comment = root_->Describe();
+  if (!models.empty()) {
+    comment =
+        absl::StrCat("models:", absl::StrJoin(models, ","), "\n", comment);
   }
 
   // Convert child visit counts to a probability distribution, pi.
+  std::array<float, kNumMoves> search_pi;
```

Comment:
is there an issue with this alloc happening every move?

Reply:
A `std::array` is allocated on the stack. There's an extra copy happening here, which is unfortunate but since it's once per move rather than once per read, I'm not that bothered.